### PR TITLE
add the config option "share-folder"

### PR DESCRIPTION
**Pull request by szaimen**
fix https://github.com/nextcloud/vm/issues/1052

### CODE REVIEW

File: nextcloud-startup-script.sh [hunk 1]
```diff
@@ -402,7 +403,13 @@ do
             clear
             run_static_script cookielifetime
         ;;
-
+		
+        "Share-Folder")
+            clear
+            occ_command config:system:set share_folder --value="/Shared"
+            msg_box "All new Nextcloud-shares from other users will be stored in the 'Shared' folder by now."
```
Comment (enoch85) [inline nextcloud-startup-script.sh]: ... users will be stored in the 'Shared' folder from now on

File: nextcloud-startup-script.sh [hunk 2]
```diff
@@ -377,7 +377,8 @@ whiptail --title "Extra configurations" --checklist --separate-output "Choose wh
 "Security" "(Add extra security based on this http://goo.gl/gEJHi7)" OFF \
 "Static IP" "(Set static IP in Ubuntu with netplan.io)" OFF \
 "Automatic updates" "(Automatically update your server every week on Sundays)" OFF \
-"CookieLifetime" "(Configure forced logout timeout for users using the web GUI)" OFF 2>results
+"CookieLifetime" "(Configure forced logout timeout for users using the web GUI)" OFF \
+"Share-folder" "(If configured, all Nextcloud-shares from other users will be stored in the 'Shared' folder)" OFF 2>results
 
 while read -r -u 9 choice
```
Comment (enoch85) [inline nextcloud-startup-script.sh]: no need to change to 10 here?

Comment (szaimen) [reply nextcloud-startup-script.sh]: I guess no: but actually I don't know what that number does.

File: nextcloud-startup-script.sh (hunk 1 above)
Comment (szaimen) [reply nextcloud-startup-script.sh]: done 👍

File: nextcloud-startup-script.sh (hunk 2 above)
Comment (enoch85) [reply nextcloud-startup-script.sh]: Me neither to be honest. :)

Comment (enoch85) [reply nextcloud-startup-script.sh]: I saw you changed it in other places though, hence the question.

File: nextcloud-startup-script.sh [hunk 3]
```diff
@@ -402,7 +403,13 @@ do
             clear
             run_static_script cookielifetime
         ;;
-
+		
+        "Share-Folder")
+            clear
+            occ_command config:system:set share_folder --value="/Shared"
```
Comment (enoch85) [inline nextcloud-startup-script.sh]: Don't we need to create the `/Shared` folder, or is it created automatically?

**Comment by enoch85** [comment]
Please be more clear on which shares are saved where in a second `msg_box` then ask if the user want's to change it with `ask_yes_no`.

File: nextcloud-startup-script.sh (hunk 3 above)
Comment (szaimen) [reply nextcloud-startup-script.sh]: It is created automatically. 👍

File: nextcloud-startup-script.sh (hunk 2 above)
Comment (szaimen) [reply nextcloud-startup-script.sh]: I changed it because I thought it was the max number of options to choose from ;)

**Comment by szaimen** [comment]
> Please be more clear on which shares are saved where in a second `msg_box` then ask if the user want's to change it with `ask_yes_no`.

this should be done 👍

File: nextcloud-startup-script.sh (hunk 2 above)
Comment (szaimen) [reply nextcloud-startup-script.sh]: Maybe we should test if removing the 9 changes anything :D

**Comment by enoch85** [comment]
OK, now I understand it better.

> Otherwise you will not be able to get shares from other Nextcloud-users.    

This makes me worried. Are you talking about `www-data` permissions? n that case, skip that part and run `setup_secure_permissions.sh` instead as a proactive action.

I can do that now.

**Comment by szaimen** [comment]
> OK, now I understand it better.
> 
> > Otherwise you will not be able to get shares from other Nextcloud-users.
> 
> This makes me worried. Are you talking about `www-data` permissions? n that case, skip that part and run `setup_secure_permissions.sh` instead as a proactive action.
> 

No, I am not talking about www-data permissions.
Your Nextcloud root folder is only not writeable, if you use the external storage app to change the root folder of the nextcloud-users and make it not-writeable.

**Comment by enoch85** [comment]
What do you think about the latest commit?

**Comment by szaimen** [comment]
> What do you think about the latest commit?

have you read my comment above?
I hope this explains everything. 
(I would uncomment `#occ_command config:system:set share_folder --value="/Shared"` and remove `bash "$SECURE"` again.)

**Comment by szaimen** [comment]
@enoch85 now it looks already better but we don't need `bash "$SECURE"` here.

**Comment by enoch85** [comment]
> Your Nextcloud root folder is only not writeable

I don't understand that.

If this has something to do with external storage's I absolutely don't think this is suited for the start-up script. It's better suited as a part of smbmount as I mentoned before.

You say "rights" which permissions are needed? Nextcloud is on `www-data`, nothing else.

> I could remove that part, if this is to much information and unnecessary.

I'm confused, imagine a regular user. :)

**Comment by szaimen** [comment]
> > Your Nextcloud root folder is only not writeable
> 
> I don't understand that.
> 
> If this has something to do with external storage's I absolutely don't think this is suited for the start-up script. It's better suited as a part of smbmount as I mentoned before.
> 
> You say "rights" which permissions are needed? Nextcloud is on `www-data`, nothing else.
> 
> > I could remove that part, if this is to much information and unnecessary.
> 
> I'm confused, imagine a regular user. :)

I will try to explain it differently:
- This PR will always work, if the external App isn't enabled!
- If the external App is enabled, this PR will always work, unless you have not created an external storage folder, that is called "/" and have set it to read only.
=> So this is a very custom setting and shouldn't be a problem for this PR.


Are you okay with that explanation?
I currently don't know how to explain it that you understand what I mean.




PS: if you create an external storage folder with the name "/" this folder becomes your new Nextcloud GUI root folder, which in this case wouldn't be writeable. So the "Shared" folder could not be automatically created.
and BTW: external storage is not only a smbshare but a path on the local filessystem or something else.

**Comment by szaimen** [comment]
Sry there was a wording issue in my writing above. I corrected it. Please read it again, if you have already read this.

**Comment by enoch85** [comment]
> If the external App is enabled, this PR will always work, unless you have not created an external storage folder, that is called "/" and have set it to read only.

OK, then we should add a check for that. Else it will fail. 

Actually, I still don't understand the need for this. If the user wants to to this, read the documentation. This seems like an edge case.

**Comment by szaimen** [comment]
> > If the external App is enabled, this PR will always work, unless you have not created an external storage folder, that is called "/" and have set it to read only.
> 
> OK, then we should add a check for that. Else it will fail.
> 
> Actually, I still don't understand the need for this. If the user wants to to this, read the documentation. This seems like an edge case.

I cannot add a check for this since the external storage app is enabled and mounted afterwards.
This is no edge case imo, but quite useful if you want to organize your root folder: it is a default setting in many Nextcloud instances that I know and tried.

Sry that I came up with this edge  error case with the external storage app, I don't think that this should prevent the PR from beeing merged.

But I would definitely remove the `bash "$SCRIPTS"/secure.sh`

**Comment by enoch85** [comment]
So what you are saying is that this will always work *until* they decide to change it themselves in the External Storages app? Aren't their change in GUI reflecting in `config.php`?

Also, this is not about *all* shares. just the shares in the External Storages app?

Also, what is root in this case, the root folder in Nextcloud GUI, or in CLI?

I don't feel confident in merging something that I can't even explain myself. (I don't use External Storages).

**Comment by enoch85** [comment]
Let's do this in 18.0.1 instead. 

I will make a small release now, and the rest of the versions this weekend.

**Comment by szaimen** [comment]
hm :/ but okay
I will prepare a detailed explanation with pictures for that

**Comment by szaimen** [comment]
This is my detailed explanation:
1. for reference: here is the nextcloud storage for the admin user:
image

if the admin user gets a share from another nextcloud user, it appears inside his root folder:
image

3. now the case that you have setuped the Shared folder before:
this is the users root folder again:
image

if this user now gets a share by another nextcloud user, it looks like this:
image
(so the "Shared" folder got created automatically)

if you open this folder, you find the shared file:
image

3. so that the case, that you have setuped an external storage empty root folder:
image

this is your new root folder than (and empty because no file is in it)
image

if you now get a share from another user, the shared folder gets created automatically again:
image

So everything works.

But if you make it read only:
image

and this is now the admins root folder (no shared folder in it)
image

if you now create a share, 
nothing appears hear, because the path is not writeable:
image

But if the Shared folder is already created there or added as additional external storage, the shares are visible again:
image

@enoch85 I hope this helps you in understanding the problem here, but I hope you get that this is really an edge case issue and shouldn't be a problem here.
Please remember: The share_folder has nothing to do with the external storage app!
It is a setting that also works with the default calibration and would never fail!
The problem is as I said before, that you cannot see your shared, files, if the root directory is not writeble and the Shared folder is non-existant.

**Comment by enoch85** [comment]
Now that's a good explanation! Put that in the docs! and then we can link it in the `msg_box`.

I'm fine with merging this, after documentation is done, which in any case will be in 18.0.1.

File: nextcloud-startup-script.sh [hunk 4]
```diff
@@ -402,7 +403,18 @@ choice=$(< results)
             clear
             run_static_script cookielifetime
         ;;
-
+		
+        "Share-folder")
```
Comment (enoch85) [inline nextcloud-startup-script.sh]: remove the " " here please.

Comment (enoch85) [reply nextcloud-startup-script.sh]: No, don't do it. It's correct.

**Comment by szaimen** [comment]
outrun by https://github.com/nextcloud/vm/pull/1070?

**Comment by szaimen** [comment]
I will close this since its outrun by https://github.com/nextcloud/vm/pull/1070.

**Review by Jegbvgiyg — COMMENTED** [review]
nextcloud-startup-script.sh[]()

**Pull request closed without merging**